### PR TITLE
emul: refactor to resolve emulators at compile time

### PR DESCRIPTION
**Pull request by gmarull**
I didn't like what I found in the emulator subsystem while looking for
places where I could apply DT_FOREACH_CHILD_SEP* macros.

This patch removes the unnecessary runtime overhead ("emul_get_binding")
by using the same tricks as regular devices do. There are probably other
issues that should be fixed in this subsystem, but that's something for
the future or for the maintainers.

### CODE REVIEW

File: include/zephyr/drivers/emul.h [hunk 1]
```diff
@@ -83,10 +81,10 @@ extern const struct emul __emul_list_end[];
  * @param data_ptr emulator-specific data
  */
 #define EMUL_DEFINE(init_ptr, node_id, cfg_ptr, data_ptr)                                          \
-	static struct emul EMUL_REG_NAME(node_id) __attribute__((__section__(".emulators")))       \
+	const struct emul EMUL_REG_NAME(node_id) __attribute__((__section__(".emulators")))        \
 	__used = {                                                                                 \
 		.init = (init_ptr),                                                                \
-		.dev_label = DT_LABEL(node_id),                                                    \
+		.name = DT_LABEL(node_id),                                                         \
```
Comment (galak) [inline include/zephyr/drivers/emul.h:87]: nit: extra whitespace at end here.

Do we need `name` at all?  If so, can we use `Z_DEVICE_DT_DEV_NAME`

**Comment by galak** [comment]
Also looks like we have https://github.com/zephyrproject-rtos/zephyr/pull/44883

File: include/zephyr/drivers/emul.h (hunk 1 above)
Comment (gmarull) [reply include/zephyr/drivers/emul.h:87]: emulators are defined independently of devices (not sure the reasons), so unless we collect it here there's no way to retrieve it (we can't do like `dev->name`).

File: drivers/spi/spi_emul.c [hunk 2]
```diff
@@ -117,13 +117,10 @@ static struct spi_driver_api spi_emul_api = {
...
 };
 
-#define EMUL_LINK_AND_COMMA(node_id) {		\
-	.label = DT_LABEL(node_id),		\
-},
-
 #define SPI_EMUL_INIT(n) \
-	static const struct emul_link_for_bus emuls_##n[] = { \
-		DT_FOREACH_CHILD(DT_DRV_INST(0), EMUL_LINK_AND_COMMA) \
+	DT_INST_FOREACH_CHILD_SEP(n, EMUL_DT_DECLARE, (;)); \
```
Comment (gmarull) [inline drivers/spi/spi_emul.c:121]: I realized I can't do `;` at the end here, it will failt if there are no children, we'll need `_TERM` versions (terminate every entry).

**Comment by alevkoy** [comment]
As you noted, part of this PR overlaps with #44883. I believe that PR is almost ready to merge (pending approval by @mbolivar-nordic). Personally, I'd prefer that this PR be rebased on top of my PR after it merges rather than merging first and creating conflicts for that PR. :)

File: include/zephyr/drivers/emul.h [hunk 3]
```diff
@@ -102,20 +100,6 @@ extern const struct emul __emul_list_end[];
...
- * linker. If the emulator structure is needed for custom API calls, it can be retrieved by the name
- * that the emulator exposes to the system (this is the devicetree node's label by default).
- *
- * @param name Emulator name to search for.  A null pointer, or a pointer to an
- * empty string, will cause NULL to be returned.
- *
- * @return pointer to emulator structure; NULL if not found or cannot be used.
- */
-const struct emul *emul_get_binding(const char *name);
-
```
Comment (yperess) [inline include/zephyr/drivers/emul.h:118]: Why is this being removed? It's very widely used downstream.

File: subsys/emul/emul_bmi160.c [hunk 4]
```diff
@@ -347,7 +347,7 @@ static int emul_bosch_bmi160_init_spi(const struct emul *emul,
 	data->emul_spi.chipsel = cfg->chipsel;
 	data->emul_spi.parent = emul;
 
-	int rc = spi_emul_register(parent, emul->dev_label, &data->emul_spi);
```
Comment (yperess) [inline subsys/emul/emul_bmi160.c:350]: This conflicts with https://github.com/zephyrproject-rtos/zephyr/pull/46818

File: include/zephyr/drivers/emul.h (hunk 1 above)
Comment (yperess) [reply include/zephyr/drivers/emul.h:87]: The reason for emulators to be defined independently is that they use the same node in the devicetree. So they would have the same label which would cause a conflict with calling `device_get_binding`. The dts will look like:

```
/ {
  &i2c0 {
    my_dev: my_dev@80 {
      ...
    };
  };
};
```

Doing this means we can use the exact same dts file for both the real board, and only need to change the compatible string of the `i2c0` bus for the host test. The device will then generate both a normal device driver AND an emulator using the same label and compatible string.

**Review by yperess — CHANGES_REQUESTED** [review]
I would much prefer to wait for https://github.com/zephyrproject-rtos/zephyr/pull/46818 and https://github.com/zephyrproject-rtos/zephyr/pull/44883 to merge as they seem to handle the majority of these concerns.

File: include/zephyr/drivers/emul.h (hunk 3 above)
Comment (gmarull) [reply include/zephyr/drivers/emul.h:118]: All emulators seem to come from DT, which means there's no case where they can't be resolved at compile time. Maybe deprecate?

**Comment by gmarull** [comment]
> As you noted, part of this PR overlaps with #44883. I believe that PR is almost ready to merge (pending approval by @mbolivar-nordic). Personally, I'd prefer that this PR be rebased on top of my PR after it merges rather than merging first and creating conflicts for that PR. :)

Yes no problem, I was not aware of #44883.

File: include/zephyr/drivers/emul.h (hunk 1 above)
Comment (gmarull) [reply include/zephyr/drivers/emul.h:87]: > The reason for emulators to be defined independently is that they use the same node in the devicetree. So they would have the same label which would cause a conflict with calling `device_get_binding`. The dts will look like:
> 
> ```
> / {
>   &i2c0 {
>     my_dev: my_dev@80 {
>       ...
>     };
>   };
> };
> ```
> 
> Doing this means we can use the exact same dts file for both the real board, and only need to change the compatible string of the `i2c0` bus for the host test. The device will then generate both a normal device driver AND an emulator using the same label and compatible string.

I still don't understand why device infrastructure can't be used (or at least re-used). Could you provide a real example (and how it is used) I can look at?

Comment (galak) [reply include/zephyr/drivers/emul.h:87]: Agreed, I'm confused by the statement of `emulators are defined independently`.  What exactly do you mean.

How or when don't we have unique devicetree nodes for the emulated devices?

Comment (galak) [reply include/zephyr/drivers/emul.h:87]: Why do `struct emul` need a `dev_label`, why can't it use `struct device *`

Comment (alevkoy) [reply include/zephyr/drivers/emul.h:87]: > Agreed, I'm confused by the statement of emulators are defined independently. What exactly do you mean.

> Why do `struct emul` need a `dev_label`, why can't it use `struct device *`

I started typing up a reply that got longer than I wanted. I'll try to be concise here, but I think I can describe this unambiguously at length. I'll avoid unqualified use of the word "device," because it seems to be overloaded in this context. I will use "component" to refer to a physical part on a board, because I don't think that word is being used for anything else here. "Emulator" also seems to be overloaded, and I don't have a great solution for that. I'll use it here to refer to a `struct emul` defined in `zephyr/subsys/emul` or the code that contains such a definition.

A `struct device` holds the runtime state for a device driver instance. In a typical configuration with devicetree, there would be one `struct device`/driver instance for each node in the tree. This is true for a production build on real hardware and for a test build with simulated hardware.

A `struct emul` holds the runtime state for an emulator instance, where an emulator instance corresponds to a physical component that would exist on a production board. In a typical test configuration using emulators, there would be `struct emul`s corresponding to several nodes in the tree. For the nodes in use, there is a `struct emul` (representing a component) and also a `struct device` (driver instance).

> I still don't understand why device infrastructure can't be used (or at least re-used). Could you provide a real example (and how it is used) I can look at?

Consider `bmi160.c`, `emul_bmi160.c`, and `tests/drivers/sensor/accel/boards/native_posix.overlay`. These are a driver for the BMI160 component, an emulator for that component, and a test that uses that emulator to exercise that driver. All 3 of these use devicetree nodes with the compatible string "bosch,bmi160". There are a `struct emul` and a `struct device` corresponding to the same DT node. AFAICT, this is the pattern in all of the existing tests in this repo that use emulators.

> I still don't understand why emulators are not regular devices. It looks like `gpio_emul.c` is, or `eeprom_emulator.c`.

These files define device drivers, but the drivers and their DT nodes don't correspond to any physical component. I'm not familiar with this code, but it looks like it's meant to be used when tests conceptually need GPIO or EEPROM, but the test isn't supposed to exercise any particular GPIO controller driver or EEPROM driver. I see a few similar examples in `bbram_emul.c` and `adc_emul.c`, which are both drivers.

These types of drivers are useful, but you can't use them to exercise the driver for a specific BBRAM or ADC component. To do that, you would need an emulator, as in a `struct emul`. You _could_ just directly simulate the I2C traffic in a driver unit test, but that gets pretty tedious in a large test for a complicated driver. `struct emul` provides a useful encapsulation for the behavior of a component.

*tl;dr:* `struct device` and `struct emul` represent distinct concepts. It's appropriate for a `struct device` and a `struct emul` to be based on the same DT node.

File: include/zephyr/drivers/emul.h (hunk 3 above)
Comment (yperess) [reply include/zephyr/drivers/emul.h:118]: Then why not deprecate `device_get_binding`? I think there can be use cases for wanting/needing to update a `const struct emul *` variable at runtime.

Comment (gmarull) [reply include/zephyr/drivers/emul.h:118]: I suggested it because it appears that all emulators can be resolved at compile time (unlike devices).

**Comment by mbolivar-nordic** [comment]
@alevkoy thanks for the great explanation in https://github.com/zephyrproject-rtos/zephyr/pull/47388#discussion_r916334794.

It's kind of unfortunate that this is "tribal knowledge" not covered in https://docs.zephyrproject.org/latest/hardware/emulator/index.html, though. The fact that @galak , @gmarull , and I were all confused and had the same questions which weren't covered there points at a gap in the documentation.

Since Google is leading the emulator effort AFAICT these days, can y'all please discuss internally how to fix this gap and update the docs?

**Comment by gmarull** [comment]
closing this as emulators rework is being done in other PRs
